internal/providers/provider.go

```diff
@@ -237,6 +237,21 @@ type ReadResourceRequest struct {
 	// each provider, and it should not be used without coordination with
 	// HashiCorp. It is considered experimental and subject to change.
 	ProviderMeta cty.Value
+
+	// DeferralAllowed signals that the provider is allowed to defer the
+	// changes. If set the caller needs to handle the deferred response.
+	DeferralAllowed bool
+}
+
+const (
+	DEFERRED_REASON_UNKNOWN                 = 0
+	DEFERRED_REASON_RESOURCE_CONFIG_UNKNOWN = 1
+	DEFERRED_REASON_PROVIDER_CONFIG_UNKNOWN = 2
+	DEFERRED_REASON_ABSENT_PREREQ           = 3
+)
+
+type Deferred struct {
+	Reason int32
 }
```

Comment:
```suggestion

type DeferredReason int32

const (
	DeferredReasonUnknown DeferredReason = iota
	DeferredReasonResourceConfigUnknown
	DeferredReasonProviderConfigUnknown
	DeferredReasonResourceConfigUnknown
	DeferredReasonAbsentPrereq
)

type Deferred struct {
	Reason DeferredReason
}
```

You can use the `iota` keyword to make this a bit easier in go. We also don't use that kind of all caps with underscores syntax in go, just camel case (is it camel case?) is fine.